### PR TITLE
EoTs of natural image corruptions

### LINKED ISSUE #852: Add EoT for image modifications like brightness, contrast, etc.
We would like to add tools for Expectation over Transformation (EoT) over image modifications like brightness, contrast, etc. in `art.preprocessing.expectation_over_transformation`. This issue should consider a general implementation or one specific for each transformation.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request adds Expectation over Transformation (EoT) preprocessing for natural image corruptions. These EoTs enable sampling multiple transformed samples in each forward pass and are fully differentiable for accurate loss gradient calculation. They can be chained together in sequence and are implemented fully framework-specific in PyTorch and TensorFlow v2. These EoTs should be useful in applications like physical adversarial patch generation. This PR includes EoTs for brightness, contrast, Gaussian noise, shot noise, and zoom blur. Additional EoTs and corruptions will be added in future PRs and releases.

Fixes #852

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/preprocessing/expectation_over_transformation/natural_corruptions/gaussian_noise/pytorch.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTBrightnessPyTorch.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param std: Range to sample the standard deviation for the Gaussian distribution. A single float
+            is translated to range [0, std]. The applied delta is sampled uniformly from this range for each image.
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/gaussian_noise/pytorch.py:53]: Please re-indent this line to align with above sentence.

File: art/preprocessing/expectation_over_transformation/natural_corruptions/gaussian_noise/tensorflow.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTGaussianNoiseTensorFlow.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param std: Range to sample the standard deviation for the Gaussian distribution. A single float
+            is translated to range [0, std]. The applied delta is sampled uniformly from this range for each image.
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/gaussian_noise/tensorflow.py:53]: Please re-indent this line.

File: art/preprocessing/expectation_over_transformation/natural_corruptions/shot_noise/pytorch.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTShotNoisePyTorch.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param lam: Range to sample the rate of the Poisson distribution. A single float
+            is translated to range [0.0, lam] or a tuple of floats is used to create sampling range
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/shot_noise/pytorch.py:53]: Please re-indent this line.

File: art/preprocessing/expectation_over_transformation/natural_corruptions/shot_noise/tensorflow.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTShotNoiseTensorFlow.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param lam: Range to sample the rate of the Poisson distribution. A single float
+            is translated to range [0.0, lam] or a tuple of floats is used to create sampling range
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/shot_noise/tensorflow.py:53]: Please re-indent this line.

File: art/preprocessing/expectation_over_transformation/natural_corruptions/zoom_blur/pytorch.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTZoomBlurPyTorch.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param zoom: Range to sample the zoom factor. A single float is translated to range [1.0, zoom] or a tuple of
+            floats is used to create sampling range [zoom[0], zoom[1]]. The applied zoom is sampled uniformly from this
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/zoom_blur/pytorch.py:53]: Please re-indent this line.

File: art/preprocessing/expectation_over_transformation/natural_corruptions/zoom_blur/tensorflow.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        apply_fit: bool = False,
+        apply_predict: bool = True,
+    ) -> None:
+        """
+        Create an instance of EoTZoomBlurTensorFlow.
+
+        :param nb_samples: Number of random samples per input sample.
+        :param clip_values: Tuple of float representing minimum and maximum values of input `(min, max)`.
+        :param zoom: Range to sample the zoom factor. A single float is translated to range [1.0, zoom] or a tuple of
+            floats is used to create sampling range [zoom[0], zoom[1]]. The applied zoom is sampled uniformly from this
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/natural_corruptions/zoom_blur/tensorflow.py:53]: Please re-indent this line.

File: art/preprocessing/expectation_over_transformation/tensorflow.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+        :param x: Input samples.
+        :param y: Label of the samples `x`.
+        :return: Transformed samples and labels.
+        """
+        raise NotImplementedError
+
+    def forward(self, x: "tf.Tensor", y: Optional["tf.Tensor"] = None) -> Tuple["tf.Tensor", Optional["tf.Tensor"]]:
+        """
+        Apply corruptions to inputs `x`.
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/tensorflow.py:70]: I think this `forward` function is not only applied for corruptions but also for image_rotation and other transformations in the future. So could you please generalize this sentence.

File: art/preprocessing/expectation_over_transformation/pytorch.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        :param y: Label of the samples `x`.
+        :return: Transformed samples and labels.
+        """
+        raise NotImplementedError
+
+    def forward(
+        self, x: "torch.Tensor", y: Optional["torch.Tensor"] = None
+    ) -> Tuple["torch.Tensor", Optional["torch.Tensor"]]:
+        """
+        Apply corruptions to inputs `x`.
```
Comment (minhitbk) [inline art/preprocessing/expectation_over_transformation/pytorch.py:72]: I think this forward function is not only applied for corruptions but also for image_rotation and other transformations in the future. So could you please generalize this sentence.

**Review by minhitbk — APPROVED** [review]
Hi @beat-buesser : Your code looks good to me and well structured. I just have a few minor comments on the doc. In addtion, I would like if possible to suggest you to add all the transformations into an __init__ file in the preprocessing folder or the like. It is because I found some import paths are very long if users want to use an eot transformation.

**Comment by beat-buesser** [comment]
Hi @minhitbk Thank you very much! I have implemented all comments.
